src/store-api/src/region_request.rs

```diff
@@ -189,9 +174,236 @@ pub struct RegionOpenRequest {
+            AlterKind::AddColumns { columns } => {
+                let mut names = HashSet::with_capacity(columns.len());
+                for col_to_add in columns {
+                    ensure!(
+                        !names.contains(&col_to_add.column_metadata.column_schema.name),
+                        InvalidRegionRequestSnafu {
+                            region_id: metadata.region_id,
+                            err: format!(
+                                "add column {} more than once",
+                                col_to_add.column_metadata.column_schema.name
+                            ),
+                        }
+                    );
+                    col_to_add.validate(metadata)?;
+                    names.insert(&col_to_add.column_metadata.column_schema.name);
+                }
+            }
+            AlterKind::DropColumns { names } => {
+                for name in names {
+                    Self::validate_column_to_drop(name, metadata)?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Returns an error if the column to drop is invalid.
+    fn validate_column_to_drop(name: &str, metadata: &RegionMetadata) -> Result<()> {
+        let column = metadata
+            .column_by_name(name)
+            .with_context(|| InvalidRegionRequestSnafu {
+                region_id: metadata.region_id,
+                err: format!("column {} does not exist", name),
+            })?;
+        ensure!(
+            column.semantic_type == SemanticType::Field,
+            InvalidRegionRequestSnafu {
+                region_id: metadata.region_id,
+                err: format!("column {} is not a field and could not be dropped", name),
+            }
+        );
+        Ok(())
+    }
+}
+
+impl TryFrom<alter_request::Kind> for AlterKind {
+    type Error = MetadataError;
+
+    fn try_from(kind: alter_request::Kind) -> Result<Self> {
+        let alter_kind = match kind {
+            alter_request::Kind::AddColumns(x) => {
+                let columns = x
+                    .add_columns
+                    .into_iter()
+                    .map(|x| x.try_into())
+                    .collect::<Result<Vec<_>>>()?;
+                AlterKind::AddColumns { columns }
+            }
+            alter_request::Kind::DropColumns(x) => {
+                let names = x.drop_columns.into_iter().map(|x| x.name).collect();
+                AlterKind::DropColumns { names }
+            }
+        };
+
+        Ok(alter_kind)
+    }
+}
+
+/// Adds a column.
+#[derive(Debug)]
+pub struct AddColumn {
+    /// Metadata of the column to add.
+    pub column_metadata: ColumnMetadata,
+    /// Location to add the column. If location is None, the region adds
+    /// the column to the last.
+    pub location: Option<AddColumnLocation>,
+}
+
+impl AddColumn {
+    /// Returns an error if the column to add is invalid.
+    pub fn validate(&self, metadata: &RegionMetadata) -> Result<()> {
```

Comment:
Need some tests for these functions:

* `AddColumn#validate`
* `AlterKind#validate`
etc.